### PR TITLE
Introduce Compaction Groups

### PR DESCRIPTION
**Pull request by raphaelsc**
Compaction group can be defined as a set of files that can be compacted together. Today, all sstables belonging to a table in a given shard belong to the same group. So we can say there's one group per table per shard. As we want to eventually allow isolation of data that shouldn't be mixed, e.g. data from different vnodes, then we want to have more than one group per table per shard. That's why compaction groups is being introduced here.

Today, all memtables and sstables are stored in a single structure per table. After compaction groups, there will be memtables and sstables for each group in the table.

As we're taking an incremental approach, table still supports a single group. But work was done on preparing table for supporting multiple groups. Completing that work is actually the next step. Also, a procedure for deriving the group from token is introduced, but today it always return the single group owned by the table. Once multiple groups are supported, then that procedure should be implemented to map a token to a group.

No semantics was changed by this series.

### CODE REVIEW

File: replica/database.hh [hunk 1]
```diff
@@ -555,7 +552,7 @@ private:
     lw_shared_ptr<sstables::sstable_set>
     do_add_sstable(lw_shared_ptr<sstables::sstable_set> sstables, sstables::shared_sstable sstable,
         enable_backlog_tracker backlog_tracker);
-    void add_sstable(sstables::shared_sstable sstable);
+    void add_sstable(compaction_group& cg, sstables::shared_sstable sstable);
```
Comment (avikivity) [inline replica/database.hh:554]: I expected to see table::add_sstable(sstable) that figures out which compaction_group to talk to, and compaction_group::add_sstable(sstable) that does the work. How does the caller of add_sstable(compaction_group, sstable) know to select the correct compaction_group? Or to split the sstable if needed?

File: replica/table.cc [hunk 2]
```diff
@@ -405,8 +407,8 @@ table::do_add_sstable(lw_shared_ptr<sstables::sstable_set> sstables, sstables::s
     return new_sstables;
 }
 
-void table::add_sstable(sstables::shared_sstable sstable) {
-    _main_sstables = do_add_sstable(_main_sstables, std::move(sstable), enable_backlog_tracker::yes);
+void table::add_sstable(compaction_group& cg, sstables::shared_sstable sstable) {
+    cg._main_sstables = do_add_sstable(cg._main_sstables, std::move(sstable), enable_backlog_tracker::yes);
```
Comment (avikivity) [inline replica/table.cc:417]: This really should delegate to compaction_group so it is control of what is happening. Maybe compaction_group shouldn't let table reach into its internals (but vice versa is okay).

File: replica/database.hh (hunk 1 above)
Comment (raphaelsc) [reply replica/database.hh:554]: table::add_sstable(compaction_group& cg, sstables::shared_sstable sstable) is a private function. I think you're referring to table::add_sstable_and_update_cache() which is in the public namespace. And yes, it should figure out which group a file belongs to. In this series, add_sstable_and_update_cache() is adding the sstable on behalf of the single group. With my next step, which is to add support to multiple groups, then a function will indeed be introduced to figure out the ownership of the sstable file.

File: replica/table.cc (hunk 2 above)
Comment (raphaelsc) [reply replica/table.cc:417]: Makes sense. table::add_sstable(compaction_group, ...) should become compaction_group::add_sstable(...). Same logic applying elsewhere.

**Comment by raphaelsc** [comment]
v2:
- table no longer have access to compaction_group private namespace
- fixed scylla-gdb memtables
- fast path in table::apply(const frozen_mutation& m, ...) if single group, to avoid unnecessary token calculation.

**Comment by raphaelsc** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1897/

Failed due to a flakiness in test/raft/randomized_nemesis_test. Already reported to raft devs.

File: replica/database.hh [hunk 3]
```diff
@@ -318,6 +318,7 @@ struct table_stats;
 using column_family_stats = table_stats;
 
 class database_sstable_write_monitor;
+class compaction_group;
```
Comment (bhalevy) [inline replica/database.hh:321]: Shouldn't it be in the sstables namespace?
(Maybe we need a compaction namespace by the way, but that's orthogonal to this change)

File: replica/database.hh [hunk 4]
```diff
@@ -409,6 +410,7 @@ private:
     lw_shared_ptr<memtable_list> make_memory_only_memtable_list();
     lw_shared_ptr<memtable_list> make_memtable_list();
 
+    compaction_manager& _compaction_manager;
```
Comment (bhalevy) [inline replica/database.hh:411]: nit: why moving _compaction_manager here is needed?

File: replica/database.hh [hunk 5]
```diff
@@ -1102,6 +1105,25 @@ public:
     void enable_off_strategy_trigger();
 
     compaction::table_state& as_table_state() const noexcept;
+
+    friend class compaction_group;
+};
+
+// Compaction group is a set of SSTables which are eligible to be compacted together.
+// By this definition, we can say:
+//      - A group contains SSTables from the same table and shard
```
Comment (bhalevy) [inline replica/database.hh:1118]: Aren't we currently serving each compaction_group by all shards?

File: replica/database.hh [hunk 6]
```diff
@@ -1102,6 +1105,25 @@ public:
...
+//      - Each group can be thought of an isolated LSM tree.
+//      - Time Window strategy has one group per window
+class compaction_group {
+    table& _t;
+    std::unique_ptr<table::table_state> _table_state;
+public:
+    compaction_group(table& t);
+
+    // Will stop ongoing compaction on behalf of this group, etc.
+    future<> stop();
```
Comment (bhalevy) [inline replica/database.hh:1133]: noexcept please.
exception in the `stop()` path are evil.

File: replica/database.hh (hunk 3 above)
Comment (bhalevy) [reply replica/database.hh:321]: Actually, I see that `table_state` is already defined in namespace compaction.
It makes sense to do the same for `compaction_group`.

File: test/boost/sstable_compaction_test.cc [hunk 7]
```diff
@@ -444,12 +444,13 @@ static future<std::vector<unsigned long>> compact_sstables(test_env& env, sstrin
...
         } else if (strategy == compaction_strategy_type::leveled) {
+            std::vector<sstables::shared_sstable> candidates;
+            candidates.reserve(sstables->size());
             for (auto& sst : *sstables) {
                 BOOST_REQUIRE(sst->get_sstable_level() == 0);
                 BOOST_REQUIRE(sst->data_size() >= min_sstable_size);
-                column_family_test(cf).add_sstable(sst);
+                candidates.push_back(sst);
             }
-            auto candidates = get_candidates_for_leveled_strategy(*cf);
```
Comment (bhalevy) [inline test/boost/sstable_compaction_test.cc:452]: let's break this into a patch of its own.
and justify it there.

File: test/boost/sstable_test.hh [hunk 8]
```diff
@@ -32,18 +32,16 @@ class column_family_test {
...
     }
 
     // NOTE: must run in a thread
     void rebuild_sstable_list(const std::vector<sstables::shared_sstable>& new_sstables,
             const std::vector<sstables::shared_sstable>& sstables_to_remove) {
-        auto permit = seastar::get_units(_cf->_sstable_set_mutation_sem, 1).get0();
-        auto builder = replica::table::sstable_list_builder(std::move(permit));
-        _cf->_main_sstables = builder.build_new_list(*_cf->_main_sstables, _cf->_compaction_strategy.make_sstable_set(_cf->schema()), new_sstables, sstables_to_remove).get0();
-        _cf->refresh_compound_sstable_set();
+        _cf->as_table_state().on_compaction_completion(sstables::compaction_completion_desc{ .old_sstables = sstables_to_remove, .new_sstables = new_sstables }, sstables::offstrategy::no).get();
```
Comment (bhalevy) [inline test/boost/sstable_test.hh:44]: This change seems intrusive enough to be separated into its own patch too.

File: test/lib/sstable_utils.hh [hunk 9]
```diff
@@ -213,6 +213,8 @@ public:
     }
 
     void set_values(sstring first_key, sstring last_key, stats_metadata stats) {
+        _sst->_data_file_size = 1;
+        _sst->_bytes_on_disk = 1;
```
Comment (bhalevy) [inline test/lib/sstable_utils.hh:217]: This too should be separated out and explained.
It seems to be valid even without compaction_groups.

File: replica/database.hh [hunk 10]
```diff
@@ -412,10 +412,10 @@ private:
 
     compaction_manager& _compaction_manager;
     sstables::compaction_strategy _compaction_strategy;
-    // SSTable set which contains all non-maintenance sstables
-    lw_shared_ptr<sstables::sstable_set> _main_sstables;
     // Holds SSTables created by maintenance operations, which need reshaping before integration into the main set
     lw_shared_ptr<sstables::sstable_set> _maintenance_sstables;
+    // Holds single compaction group owned by table, meaning that all SSTables can be compacted with one another.
+    std::unique_ptr<compaction_group> _compaction_group;
```
Comment (bhalevy) [inline replica/database.hh:414]: I suggest placing the member at its final position already in the first patch that introduced it
to reduce the churn.

Right after `_compaction_strategy` seems even more appropriate.

File: replica/database.hh [hunk 11]
```diff
@@ -1116,13 +1115,24 @@ public:
...
     future<> stop();
 
+    // Clear sstable set content, etc.
+    void clear();
+
+    // Add sstable to main set
+    void add_sstable(sstables::shared_sstable sstable);
+
+    const lw_shared_ptr<sstables::sstable_set>& main_sstables() const;
+    void update_main_sstables(lw_shared_ptr<sstables::sstable_set> new_set);
```
Comment (bhalevy) [inline replica/database.hh:1150]: nit: rename to `set_main_sstables`. `update` may imply the new_set is added into the existing set rather than replacing it.

File: replica/table.cc [hunk 12]
```diff
@@ -405,9 +407,17 @@ table::do_add_sstable(lw_shared_ptr<sstables::sstable_set> sstables, sstables::s
...
 
-void table::add_sstable(sstables::shared_sstable sstable) {
-    _main_sstables = do_add_sstable(_main_sstables, std::move(sstable), enable_backlog_tracker::yes);
-    refresh_compound_sstable_set();
+void compaction_group::add_sstable(sstables::shared_sstable sstable) {
+    _main_sstables = _t.do_add_sstable(_main_sstables, std::move(sstable), enable_backlog_tracker::yes);
+    _t.refresh_compound_sstable_set();
+}
+
+const lw_shared_ptr<sstables::sstable_set>& compaction_group::main_sstables() const {
```
Comment (bhalevy) [inline replica/table.cc:421]: nit: can be noexcept

File: replica/table.cc [hunk 13]
```diff
@@ -405,9 +407,17 @@ table::do_add_sstable(lw_shared_ptr<sstables::sstable_set> sstables, sstables::s
     return new_sstables;
 }
 
-void table::add_sstable(sstables::shared_sstable sstable) {
-    _main_sstables = do_add_sstable(_main_sstables, std::move(sstable), enable_backlog_tracker::yes);
-    refresh_compound_sstable_set();
+void compaction_group::add_sstable(sstables::shared_sstable sstable) {
+    _main_sstables = _t.do_add_sstable(_main_sstables, std::move(sstable), enable_backlog_tracker::yes);
+    _t.refresh_compound_sstable_set();
```
Comment (bhalevy) [inline replica/table.cc:418]: This violates layering.
If that's (hopefully) temporary just for doing an incremental conversion of the code,
let's add a FIXME comment explaining and documenting it.

File: replica/table.cc [hunk 14]
```diff
@@ -954,7 +971,7 @@ table::update_sstable_lists_on_off_strategy_completion(sstables::compaction_comp
 
 // Note: must run in a seastar thread
 void
-table::on_compaction_completion(sstables::compaction_completion_desc desc) {
+table::on_compaction_completion(compaction_group& cg, sstables::compaction_completion_desc desc) {
```
Comment (bhalevy) [inline replica/table.cc:994]: My gut feeling is that this method should move to compaction_group
and that should interact with the table layer to execute the table-wide code.

File: replica/database.hh [hunk 15]
```diff
@@ -1129,10 +1127,15 @@ public:
...
     void add_sstable(sstables::shared_sstable sstable);
+    // Add sstable to maintenance set
+    void add_maintenance_sstable(sstables::shared_sstable sst);
 
     const lw_shared_ptr<sstables::sstable_set>& main_sstables() const;
     void update_main_sstables(lw_shared_ptr<sstables::sstable_set> new_set);
 
+    const lw_shared_ptr<sstables::sstable_set>& maintenance_sstables() const;
+    void update_maintenance_sstables(lw_shared_ptr<sstables::sstable_set> new_set);
+
```
Comment (bhalevy) [inline replica/database.hh:1154]: ditto, rename to `set_maintenance_sstables`

File: replica/table.cc [hunk 16]
```diff
@@ -420,9 +420,17 @@ void compaction_group::update_main_sstables(lw_shared_ptr<sstables::sstable_set>
     _main_sstables = std::move(new_set);
 }
 
-void table::add_maintenance_sstable(sstables::shared_sstable sst) {
-    _maintenance_sstables = do_add_sstable(_maintenance_sstables, std::move(sst), enable_backlog_tracker::no);
-    refresh_compound_sstable_set();
+void compaction_group::add_maintenance_sstable(sstables::shared_sstable sst) {
+    _maintenance_sstables = _t.do_add_sstable(_maintenance_sstables, std::move(sst), enable_backlog_tracker::no);
+    _t.refresh_compound_sstable_set();
```
Comment (bhalevy) [inline replica/table.cc:431]: ditto FIXME

File: replica/table.cc [hunk 17]
```diff
@@ -420,9 +420,17 @@ void compaction_group::update_main_sstables(lw_shared_ptr<sstables::sstable_set>
...
+}
+
+const lw_shared_ptr<sstables::sstable_set>& compaction_group::maintenance_sstables() const {
```
Comment (bhalevy) [inline replica/table.cc:434]: ditto noexcept :)

File: replica/database.hh (hunk 6 above)
Comment (raphaelsc) [reply replica/database.hh:1133]: stop() does flush a memtable, shouldn't we fail gracefully rather than aborting?

File: replica/database.hh (hunk 4 above)
Comment (raphaelsc) [reply replica/database.hh:411]: That's because table::_compaction_manager should be set before group accesses it.

File: replica/database.hh (hunk 3 above)
Comment (raphaelsc) [reply replica/database.hh:321]: The idea is that table_state becomes compaction_group_state. compaction_group is a replica entity, rather than one which lives in compaction namespace. We did tons of effort in compaction layer to not reference table directly, but rather work with the table_state. compaction group stores memtable_list and sstable_set's, all of which belong to the replica namespace.

I was thinking yesterday about your question, and perhaps I should create replica/compaction_group.hh and replica/compaction_group.cc, which allows for better organization of table and compaction group code.

Sounds good to you?

File: test/boost/sstable_compaction_test.cc (hunk 7 above)
Comment (raphaelsc) [reply test/boost/sstable_compaction_test.cc:452]: That change was required as part of the change, as that unit test doesn't run in a thread, and get_candidates_for_leveled_strategy() would fail into is_running_in_a_thread assertion.

Comment (raphaelsc) [reply test/boost/sstable_compaction_test.cc:452]: I can move this change into a separate patch that comes before the change in column_family_test.

File: test/boost/sstable_test.hh (hunk 8 above)
Comment (raphaelsc) [reply test/boost/sstable_test.hh:44]: All the changes in "test: compaction: Prepare tests for compaction group changes" were done as a consequence of these change in column_family_test. That's because those functions now have to run in a thread. But I can prepare the tests for the preparation of the tests :-)

File: test/lib/sstable_utils.hh (hunk 9 above)
Comment (raphaelsc) [reply test/lib/sstable_utils.hh:217]: Yes, they can go into a separate patch. thanks

File: replica/database.hh (hunk 10 above)
Comment (raphaelsc) [reply replica/database.hh:414]: I probably needed to put it after maintenance set because compaction_group ctor accessed table::_maintanence_sstables and found it uninitialized. But by the end of the series, _compaction_group will reach that location as those fields are moved into compaction group class.

File: replica/database.hh (hunk 11 above)
Comment (raphaelsc) [reply replica/database.hh:1150]: Makes sense. thanks!

File: replica/table.cc (hunk 12 above)
Comment (raphaelsc) [reply replica/table.cc:421]: will add

File: replica/table.cc (hunk 13 above)
Comment (raphaelsc) [reply replica/table.cc:418]: Yes, I really dislike this too. When writing this series, I tried to limit the amount of changes, and this was one of them. But I think I can move _t.refresh_compound_sstable_set() to be called by table itself, rather than relying on compaction group to do so.

File: replica/table.cc (hunk 14 above)
Comment (raphaelsc) [reply replica/table.cc:994]: Shouldn't it be table coordinating the activity rather than the other way around? 

the update of sstable set happens through row_cache::invalidate(external_updater, ...) method, and we have table then refreshing its cache snapshot, updating its statistics, submitting files for deletion, etc.

Maybe sstable_list_updater (which implements the actual replacement), which is defined in table::on_compaction_completion() should be moved to compaction_group. Maybe it's not needed.

File: replica/database.hh (hunk 15 above)
Comment (raphaelsc) [reply replica/database.hh:1154]: Will do.

File: replica/database.hh (hunk 3 above)
Comment (bhalevy) [reply replica/database.hh:321]: The organization makes sense.
Unfortunately the name doesn't.
I think this entity is more than about compaction so it's name is misleading.
For example this group of tokens will be a unit for replication purposes, and consequently, for repair.  Eventually, it will become a tablet.

Comment (raphaelsc) [reply replica/database.hh:321]: "subtable (plural subtables) A table of data corresponding to part of a greater parent table." ? :-) 

table_state can become compaction_group (makes sense in the compaction layer), but you're right compaction_group is not a good name for that class, living in replica namespace, storing a subset of a table.

File: replica/database.hh (hunk 6 above)
Comment (bhalevy) [reply replica/database.hh:1133]: Even if flush may fail gracefully, we should return that as an exceptional future.

File: test/boost/sstable_compaction_test.cc (hunk 7 above)
Comment (bhalevy) [reply test/boost/sstable_compaction_test.cc:452]: thanks

File: replica/database.hh (hunk 4 above)
Comment (bhalevy) [reply replica/database.hh:411]: ok

File: replica/database.hh (hunk 10 above)
Comment (bhalevy) [reply replica/database.hh:414]: ok, maybe add a FIXME comment to set the stage for that

File: replica/table.cc (hunk 14 above)
Comment (bhalevy) [reply replica/table.cc:994]: I think that compaction_manager should deal with compaction_group_state that should interact with the compaction_group object, and if that needs interaction with the table, so be it.
But bypassing that and calling the table with a refwrence to the compaction group seems like violating layering.

File: replica/database.hh (hunk 3 above)
Comment (raphaelsc) [reply replica/database.hh:321]: Compaction group could still be a good name if we think that compaction group refers to a LSM storage tree that is isolated and can be operated independently from others which belong to the same table. Then naturally compaction group interface will expose method for interacting with its memtable (which is in first level of the tree) and sstable set(s)  (which belong to higher levels).

Comment (bhalevy) [reply replica/database.hh:321]: Maybe they can be already be called a tablet or simple_tablet. @avikivity what do you think?

Comment (bhalevy) [reply replica/database.hh:321]: Or would it be appropriate only with tablet partitioning / replication mapping?

Comment (bhalevy) [reply replica/database.hh:321]: Let's not hold our horses on naming.
I don't object to proceeding with `compaction_group`

Comment (avikivity) [reply replica/database.hh:321]: Let's go with compaction_group. We can always rename it later.

File: replica/compaction_group.hh [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+#pragma once
+
+namespace replica {
+
+// Compaction group is a set of SSTables which are eligible to be compacted together.
+// By this definition, we can say:
+//      - A group contains SSTables from the same table and shard
+//      - Each group can be thought of an isolated LSM tree.
+//      - Time Window strategy has one group per window
```
Comment (bhalevy) [inline replica/compaction_group.hh:20]: Doesn't that conflict with major compaction?

File: replica/compaction_group.hh [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+    lw_shared_ptr<memtable_list> _memtables;
+    // SSTable set which contains all non-maintenance sstables
+    lw_shared_ptr<sstables::sstable_set> _main_sstables;
+    // Holds SSTables created by maintenance operations, which need reshaping before integration into the main set
+    lw_shared_ptr<sstables::sstable_set> _maintenance_sstables;
+public:
+    compaction_group(table& t);
+
+    // Will stop ongoing compaction on behalf of this group, etc.
+    future<> stop();
```
Comment (bhalevy) [inline replica/compaction_group.hh:34]: noexcept please

**Comment by raphaelsc** [comment]
v3:
- rebased
- split preparatory changes for unit tests
- moved compaction_group into its dedicated header
- added noexcept qualified to methods retrieving sstable set in compaction_group
- moved functions called on compaction completion to group
- moved min_memtable_timestamp to group

File: replica/compaction_group.hh (hunk 18 above)
Comment (raphaelsc) [reply replica/compaction_group.hh:20]: No, because there's no token overlapping among groups, so we can serialize major across groups

File: replica/compaction_group.hh (hunk 19 above)
Comment (bhalevy) [reply replica/compaction_group.hh:34]: So we can/should make flush noexcept too, for exactly that reason.
It should be a very rare reason to allow an async function to fail in 2 ways - throwing an exception, and returning a failed future, because this can double the error handling efforts (in the outermost calls.  if a function throws in a continuation we convert that into an exceptional future).

The most common technique to use is:
```c++
future<> compaction_group::stop() {
  try {
    return _memtables->flush().finally([this] {
        return _t._compaction_manager.remove(as_table_state());
    });
  } catch (...) {
    return current_exception_as_future<>();
  }
}
```

This will not terminate.

Another way to do it is:
`futurize_invoke([this] { return _memtables->flush(); }).finally( ... )`

but it's cumbersome and ugly in this case.

So I'd vote for making flush noexcept to start with, possibly catching any exceptions there as described above.

Comment (raphaelsc) [reply replica/compaction_group.hh:34]: That's orthogonal to this series, I think. I think I can start with your suggestion to convert any thrown exception into exception future and add a FIXME for the subsequent work.

Comment (raphaelsc) [reply replica/compaction_group.hh:34]: That = make flush noexcept.

File: replica/database.hh [hunk 20]
```diff
@@ -729,7 +726,9 @@ public:
     // FIXME: in case a query is satisfied from a single memtable, avoid a copy
     using const_mutation_partition_ptr = std::unique_ptr<const mutation_partition>;
     using const_row_ptr = std::unique_ptr<const row>;
-    memtable& active_memtable() { return _memtables->active_memtable(); }
+    // FIXME: for supporting multiple compaction groups, this interface will need to return
+    // as many active sstables are there are groups.
```
[the remark this answers is not in the document]
Comment (raphaelsc) [reply replica/database.hh:730]: > nit: s/are there/as there/

thanks

**Comment by raphaelsc** [comment]
v4:
- fixed grammar mistake in comment
- made compaction_group::stop() noexcept
- added patch to restore indentation in table::stop()

File: test/boost/sstable_test.hh [hunk 21]
```diff
@@ -32,18 +32,16 @@ class column_family_test {
 public:
     column_family_test(lw_shared_ptr<replica::column_family> cf) : _cf(cf) {}
 
+    // NOTE: must run in a thread.
     void add_sstable(sstables::shared_sstable sstable) {
-        _cf->_main_sstables->insert(std::move(sstable));
-        _cf->refresh_compound_sstable_set();
+        auto new_sstables = { sstable };
+        _cf->as_table_state().on_compaction_completion(sstables::compaction_completion_desc{ .new_sstables = new_sstables }, sstables::offstrategy::no).get();
```
Comment (bhalevy) [inline test/boost/sstable_test.hh:38]: wouldn't it be simpler to return a `future<>` here and let the caller `.get()` it if it's in a thread?

File: replica/compaction_group.hh (hunk 18 above)
Comment (bhalevy) [reply replica/compaction_group.hh:25]: > there's no token overlapping among groups

How come? Don't all groups in TWCS cover the whole token range and we just segregate the rows across groups based on the time window?

File: replica/database.hh [hunk 22]
```diff
@@ -770,10 +774,15 @@ public:
     // Applies given mutation to this column family
     // The mutation is always upgraded to current schema.
     void apply(const frozen_mutation& m, const schema_ptr& m_schema, db::rp_handle&& h = {}) {
-        do_apply(std::move(h), m, m_schema);
+        if (auto cg = single_compaction_group_if_available()) {
+            do_apply(*cg, std::move(h), m, m_schema);
+            return;
+        }
+        dht::token t = dht::get_token(*m_schema, m.key());
+        do_apply(compaction_group_for_token(t), std::move(h), m, m_schema);
```
Comment (bhalevy) [inline replica/database.hh:782]: We can have a variant of `compaction_group_for_token` taking references to the schema and key that will hide the complexity of `single_compaction_group_if_available`, doing essentially what this function is doing, just reusable in further use cases that may need it.

File: replica/database.hh [hunk 23]
```diff
@@ -402,19 +403,17 @@ private:
...
 
+    compaction_manager& _compaction_manager;
     sstables::compaction_strategy _compaction_strategy;
-    // SSTable set which contains all non-maintenance sstables
-    lw_shared_ptr<sstables::sstable_set> _main_sstables;
-    // Holds SSTables created by maintenance operations, which need reshaping before integration into the main set
-    lw_shared_ptr<sstables::sstable_set> _maintenance_sstables;
-    // Compound set which manages all the SSTable sets (e.g. main, etc) and allow their operations to be combined
+    // TODO: Still holds a single compaction group, meaning all sstables are eligible to be compacted with one another. Soon, a table
+    //  will be able to hold more than one group.
```
Comment (bhalevy) [inline replica/database.hh:414]: How about keeping a map of compaction_group per token range
and populate it with a single entry for the full range?

This will show the basic requirements of supporting multiple groups upfront
and will reduce the churn of supporting multiple groups later on.

File: replica/compaction_group.hh (hunk 18 above)
Comment (raphaelsc) [reply replica/compaction_group.hh:25]: Ah you're talking about TWCS. That was just a conceptual statement. Will remove it to not confuse anybody else.

File: replica/database.hh (hunk 23 above)
Comment (avikivity) [reply replica/database.hh:414]: I think it's okay to start with unique_ptr. Since there's just one now all the code to locate the correct group will be buggy.

File: replica/table.cc [hunk 24]
```diff
@@ -1216,6 +1217,21 @@ table::make_memtable_list() {
     return make_lw_shared<memtable_list>(std::move(seal), std::move(get_schema), _config.dirty_memory_manager, _stats, _config.memory_compaction_scheduling_group);
 }
 
+compaction_group::compaction_group(table& t)
+    : _t(t)
+    , _table_state(std::make_unique<table_state>(t))
```
Comment (avikivity) [inline replica/table.cc]: Where is compaction_group::table_state defined?

File: replica/table.cc [hunk 25]
```diff
@@ -1216,6 +1217,21 @@ table::make_memtable_list() {
...
+{
+    _t._compaction_manager.add(as_table_state());
+}
+
+future<> compaction_group::stop() noexcept {
+    try {
+        return _t._compaction_manager.remove(as_table_state());
```
Comment (avikivity) [inline replica/table.cc]: Better to change compaction_manager::remove to be noexcept.

File: replica/table.cc [hunk 26]
```diff
@@ -2379,7 +2394,7 @@ table::as_mutation_source_excluding(std::vector<sstables::shared_sstable>& ssts)
     });
 }
 
-class table::table_state : public compaction::table_state {
+class compaction_group::table_state : public compaction::table_state {
```
Comment (avikivity) [inline replica/table.cc:2481]: How did it compile if it's defined after it's used?

**Comment by avikivity** [comment]
Looks good.

File: replica/table.cc (hunk 24 above)
Comment (raphaelsc) [reply replica/table.cc]: Justed moved table::table_state to compaction group, so it's still defined in replica/table.cc while being renamed. One of the reasons is that it makes sense, and another reason is to avoid including database.hh as I cannot forward declare a nested class type.

File: replica/table.cc (hunk 25 above)
Comment (raphaelsc) [reply replica/table.cc]: Will do.

File: replica/table.cc (hunk 26 above)
Comment (raphaelsc) [reply replica/table.cc:2481]: Both GCC and clang allow me to instantiate the *ctor* of class that wasn't defined yet. Had an member been accessed instead, then the compiler would yell at me.

This is what we have:
https://godbolt.org/z/4eTs8n73P

Comment (raphaelsc) [reply replica/table.cc:2481]: With make_unique<T>(args...), we satisfy the compiler, then the class is found defined by the time the template is evaluated.

Comment (raphaelsc) [reply replica/table.cc:2481]: Interesting reading about this topic: https://howardhinnant.github.io/incomplete.html

File: replica/database.hh (hunk 22 above)
Comment (raphaelsc) [reply replica/database.hh:782]: Makes sense. Can be named compaction_group_for_key().

**Comment by raphaelsc** [comment]
v5:
- made compaction_manager::remove() noexcept
- introduced table::compaction_group_for_key() to avoid the duplication of fast path (when table owns a single group)

File: compaction/compaction_manager.cc [hunk 27]
```diff
@@ -789,7 +789,9 @@ future<> compaction_manager::stop_tasks(std::vector<shared_ptr<task>> tasks, sst
     });
 }
 
-future<> compaction_manager::stop_ongoing_compactions(sstring reason, compaction::table_state* t, std::optional<sstables::compaction_type> type_opt) {
+future<> compaction_manager::stop_ongoing_compactions(sstring reason, compaction::table_state* t, std::optional<sstables::compaction_type> type_opt) noexcept {
+  // FIXME: restore indentation
```
Comment (bhalevy) [inline compaction/compaction_manager.cc:793]: let's either fix the indentation in this series, or drop the comment, if we can live with half-indent here :)

File: replica/compaction_group.hh (hunk 18 above)
Comment (bhalevy) [reply replica/compaction_group.hh:25]: Looks like the confusing comment is still here.
I don't know what you mean by "Time Window strategy" here, other than TWCS.

File: replica/table.cc [hunk 28]
```diff
@@ -803,17 +862,14 @@ table::stop() {
...
-                            _cache.refresh_snapshot();
-                        });
+            return _compaction_group->stop().then([this] {
+                return _sstable_deletion_gate.close().then([this] {
+                    return get_row_cache().invalidate(row_cache::external_updater([this] {
+                        _compaction_group->clear_sstables();
+                        _sstables = make_compound_sstable_set();
+                    })).then([this] {
+                        _compaction_strategy.get_backlog_tracker().disable();
+                        _cache.refresh_snapshot();
```
Comment (bhalevy) [inline replica/table.cc:872]: nit: I think that if make this a coroutine as a preliminary change, it will be more obvious what exactly changed by reducing the churn and leaving only the parts still relevant to table and what moves to compaction_group.

**Review by bhalevy — APPROVED** [review]
looks mostly good to me.
I had only a few minor comments

File: replica/table.cc (hunk 26 above)
Comment (avikivity) [reply replica/table.cc:2484]: ok

File: replica/table.cc (hunk 28 above)
Comment (raphaelsc) [reply replica/table.cc:872]: In the first patch which touches table::stop(), I used that technique of breaking indentation to make it easy to verify what changed, then restored indentation in a later patch. So I assume that making it a couroutine is not needed for this specific reason. But coroutine has its benefits so I will convert table::stop() into coroutine in preliminary patch.

File: compaction/compaction_manager.cc (hunk 27 above)
Comment (raphaelsc) [reply compaction/compaction_manager.cc:793]: Will add a patch to restore indentation.

**Review by avikivity — APPROVED** [review]
Please add before/after perf_simple_query. I don't expect surprises, but this touches the hot path.

**Comment by raphaelsc** [comment]
v6:
- Added patch to restore indentation in compaction_manager::stop_ongoing_compactions()
- Converted table::stop() into coroutine in a preliminary patch
- Improved documentation of compaction_group

**Comment by raphaelsc** [comment]
> Please add before/after perf_simple_query. I don't expect surprises, but this touches the hot path.

before:

median 951913.28 tps ( 72.2 allocs/op,  14.3 tasks/op,   58052 insns/op,        0 errors)
median 958138.97 tps ( 72.2 allocs/op,  14.3 tasks/op,   57940 insns/op,        0 errors)
median 933679.25 tps ( 72.3 allocs/op,  14.4 tasks/op,   58212 insns/op,        0 errors)
median 905033.97 tps ( 72.3 allocs/op,  14.4 tasks/op,   58633 insns/op,        0 errors)

after:

median 945253.73 tps ( 72.2 allocs/op,  14.3 tasks/op,   57867 insns/op,        0 errors)
median 945324.68 tps ( 72.2 allocs/op,  14.3 tasks/op,   58086 insns/op,        0 errors)
median 945992.10 tps ( 72.2 allocs/op,  14.3 tasks/op,   58004 insns/op,        0 errors)
median 939955.27 tps ( 72.3 allocs/op,  14.3 tasks/op,   57971 insns/op,        0 errors)

No surprise

**Comment by raphaelsc** [comment]
@bhalevy this will probably conflict with ur series to deglobalize repair history, so I'll rebase this series once yours get into master.

**Comment by raphaelsc** [comment]
@scylladb/scylla-maint ping merge.
